File: web/src/types/health.ts
import { v4 as uuidv4 } from 'uuid';

export enum HealthStatusType {
  Ok ,
  Warning ,
  Critical ,
  Degraded ,
  ShutdownPending ,
}

export enum AlertSeverity {
  Info ,
  Warning ,
  Critical,
}

export enum ActionType {
  ThrottleCameraFps ,
  DisableMlModel ,
  ReduceSensorRate ,
  RotateWalEarly ,
  DropCameraFrames ,
  ReduceLogLevel ,
  RebootSystem 
}

export interface HealthStatus {
  id: string;
  truck_id: string;
  timestamp: string;
  status: HealthStatusType;
  resources: ResourceUsage;
  tasks: TaskStatus[];
  alerts: HealthAlert[];
  actions_taken: HealthAction[];
  meta :HealthEventMetadata;
  created_at: string;
}

export interface ResourceUsage {
  cpu_percent: number;
  cpu_cores: number;
  memory_percent: number;
  memory_used_mb: number;
  memory_total_mb: number;
  memory_available_mb: number;
  swap_percent: number;
  disk_percent: number;
  disk_used_gb: number;
  disk_total_gb: number;
  disk_available_gb: number;
  temperature_c: number;
  thermal_throttling: boolean;
  uptime_sec: number;
  load_average: [number, number, number];
}

export interface TaskStatus {
  name: string;
  is_alive: boolean;
  last_seen_ms: number;
  cpu_usage_percent: number;
  memory_usage_mb: number;
  restarts: number;
  last_restart: string | null;
}

export interface HealthAlert {
  alert_id: string;
  alert_type: string;
  severity: AlertSeverity;
  message: string;
  triggered_at: string;
  source: string;
  recommended_action: string;
}

export interface HealthAction {
  action_id: string;
  action_type: ActionType;
  target_module: string;
  parameters: any; // JSON object with action-specific parameters
  executed_at: string;
  success: boolean;
  message: string;
}

export interface HealthEventMetadata {
  device_id: string;
  version: string;
  hostname: string;
  ip_address: string;
  mac_address: string;
  location: [number, number] | null;
  hardware_model: string;
}

export interface HealthSummary {
  truck_id: string;
  last_timestamp: string;
  status: HealthStatusType;
  cpu_percent: number;
  memory_percent: number;
  disk_percent: number;
  temperature_c: number;
  uptime_sec: number;
  active_alerts: number;
  health_score: number;
}

export interface HealthStats {
  total_health_events: number;
  by_status: Record<string, number>;
  by_truck: Record<string, number>;
  avg_cpu_percent: number;
  avg_memory_percent: number;
  avg_disk_percent: number;
  avg_temperature_c: number;
  last_24_hours: number;
  last_7_days: number;
  last_30_days: number;
}

export interface DashboardHealthStatus {
  recent_health_status: HealthSummary[];
  stats: HealthStats;
  status_by_type: Record<string, number>;
  top_trucks: [string, number][];
  avg_resources: AverageResources;
  time_range_hours: number;
}

export interface AverageResources {
  avg_cpu_percent: number;
  avg_memory_percent: number;
  avg_disk_percent: number;
  avg_temperature_c: number;
}

File: truck-central-server/src/models/health.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use chrono::{DateTime, Utc};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HealthStatus {
    pub id: Uuid,
    pub truck_id: Uuid,
    pub timestamp: DateTime<Utc>,
    pub status: HealthStatusType,
    pub resources: ResourceUsage,
    pub tasks: Vec<TaskStatus>,
    pub alerts: Vec<HealthAlert>,
    pub actions_taken: Vec<HealthAction>,
    pub meta HealthEventMetadata,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum HealthStatusType {
    Ok,
    Warning,
    Critical,
    Degraded,
    ShutdownPending,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceUsage {
    pub cpu_percent: f32,
    pub cpu_cores: usize,
    pub memory_percent: f32,
    pub memory_used_mb: u64,
    pub memory_total_mb: u64,
    pub memory_available_mb: u64,
    pub swap_percent: f32,
    pub disk_percent: f32,
    pub disk_used_gb: u64,
    pub disk_total_gb: u64,
    pub disk_available_gb: u64,
    pub temperature_c: f32,
    pub thermal_throttling: bool,
    pub uptime_sec: u64,
    pub load_average: (f32, f32, f32),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskStatus {
    pub name: String,
    pub is_alive: bool,
    pub last_seen_ms: u64,
    pub cpu_usage_percent: f32,
    pub memory_usage_mb: u64,
    pub restarts: u32,
    pub last_restart: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HealthAlert {
    pub alert_id: String,
    pub alert_type: String,
    pub severity: AlertSeverity,
    pub message: String,
    pub triggered_at: DateTime<Utc>,
    pub source: String,
    pub recommended_action: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum AlertSeverity {
    Info,
    Warning,
    Critical,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HealthAction {
    pub action_id: String,
    pub action_type: ActionType,
    pub target_module: String,
    pub parameters: serde_json::Value,
    pub executed_at: DateTime<Utc>,
    pub success: bool,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum ActionType {
    ThrottleCameraFps,
    DisableMlModel,
    ReduceSensorRate,
    RotateWalEarly,
    DropCameraFrames,
    ReduceLogLevel,
    RebootSystem,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HealthEventMetadata {
    pub device_id: String,
    pub version: String,
    pub hostname: String,
    pub ip_address: String,
    pub mac_address: String,
    pub location: Option<(f64, f64)>,
    pub hardware_model: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HealthSummary {
    pub truck_id: Uuid,
    pub last_timestamp: DateTime<Utc>,
    pub status: HealthStatusType,
    pub cpu_percent: f32,
    pub memory_percent: f32,
    pub disk_percent: f32,
    pub temperature_c: f32,
    pub uptime_sec: u64,
    pub active_alerts: i32,
    pub health_score: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HealthStats {
    pub total_health_events: i64,
    pub by_status: std::collections::HashMap<String, i64>,
    pub by_truck: std::collections::HashMap<Uuid, i64>,
    pub avg_cpu_percent: f32,
    pub avg_memory_percent: f32,
    pub avg_disk_percent: f32,
    pub avg_temperature_c: f32,
    pub last_24_hours: i64,
    pub last_7_days: i64,
    pub last_30_days: i64,
}